File: apps/android-frontend/android/app/src/main/java/au/com/codecapers/photosphere/jsengine/SyncDriver.java
package au.com.codecapers.photosphere.jsengine;

//
// The background sync: one pass at a time, and a loop above it.
//
// A pass asks the plan-sync task whether a sync should run and runs the tasks it hands back, in
// order. The loop asks for a pass, waits for it, waits again for the gap the plan asked for, and asks
// for another, until the driver is stopped.
//
// Unlike the import loop, nothing a plan says ends this one. Every reason a sync is refused can go
// away without the app being touched: a phone moves onto Wi-Fi, a network comes back, a database gets
// an origin, the user switches syncing back on. A loop that ended on a refusal would need something
// to notice each of those and start it again, and the cost of getting that wrong is a phone that
// looks like it is backing up and is not. A refused pass costs a settings file read.
//
// There is exactly one entry point that runs a pass, and it is serialised: asked to run while a pass
// is in flight, it waits for that pass and returns its outcome rather than starting a second. On
// Android only the loop asks; on iOS the foreground loop and the system's background task both ask,
// and neither knows about the other, because they do not have to.
//
// A pass also takes the shared BackgroundPassLock, so a sync and an import never run at once. A pass
// that cannot take it is skipped and tried again after the usual gap.
//
// Nothing in this class touches Android, which is why the decisions it makes can be unit tested. The
// service supplies everything that does through the Host interface below.
//
public final class SyncDriver {

    //
    // The name this driver takes the shared pass lock under, so a skipped pass can say what it is
    // waiting behind.
    //
    public static final String PASS_LOCK_OWNER = "sync";

    //
    // The gap between passes when no plan has said what it should be, in milliseconds.
    //
    // Only reached when the very first plan read fails, because every plan carries a gap that the
    // settings file resolves (see resolveSyncPauseMs in packages/api/src/lib/sync-settings.ts, which
    // is where the value a running phone uses comes from). It exists because a gap of zero is not a
    // gap: the loop would ask for a pass, fail to read a plan, and ask again as fast as the engine
    // could answer.
    //
    private static final long FALLBACK_PAUSE_MS = 5 * 60 * 1000;

    //
    // What a finished pass says about what should happen next.
    //
    public enum PassOutcome {

        //
        // The pass ran, was refused, or tried and failed. Wait the gap and go again. There is
        // deliberately no other value: nothing a pass discovers is a reason to stop syncing for good.
        //
        RAN,
    }

    //
    // Everything the driver needs that it cannot do itself: talking to the engine pool, waiting,
    // keeping the CPU awake, and saying what happened.
    //
    // It is an interface so the loop's decisions can be tested without an Android device, a running
    // engine, or a real clock.
    //
    public interface Host {

        //
        // Runs the plan-sync task and returns what it says a pass should do. Throws when the task
        // fails, which the driver treats as a pass that did not work rather than as a reason to give
        // up.
        //
        SyncPlan readPlan() throws Exception;

        //
        // Runs one of the plan's steps and waits for it to finish. Returns true when it succeeded.
        //
        boolean runStep(SyncPlan.Step step) throws Exception;

        //
        // Waits for the given number of milliseconds, or until the driver is stopped. Returns false
        // when the wait was cut short by a stop, so the loop ends instead of starting another pass.
        //
        boolean pause(long millis) throws InterruptedException;

        //
        // Keeps the CPU running while a pass is in flight, and lets it sleep again afterwards.
        //
        // Held for the length of a pass rather than the life of the service, because a foreground
        // service keeps the process alive but does not keep the CPU awake once the screen is off,
        // and a wake lock held all night flattens the phone. A sync pass is shorter than an import
        // pass, but a sync of a large backlog over a slow connection is not short.
        //
        void holdAwake(boolean awake);

        //
        // Reports what the background sync is doing. Goes to logcat on a device: the app log is
        // written over a socket from the WebView, which is suspended exactly when this matters.
        //
        void report(String message);

        //
        // Reports something that went wrong.
        //
        void reportError(String message);
    }

    //
    // Everything the driver needs that it cannot do itself.
    //
    private final Host host;

    //
    // The lock that keeps a sync pass and an import pass apart.
    //
    private final BackgroundPassLock sharedPassLock;

    //
    // Guards the pass bookkeeping below, and is what a second caller waits on while a pass runs.
    //
    private final Object passLock = new Object();

    //
    // True while a pass is in flight, so a second request waits for it rather than starting another.
    //
    private boolean passRunning = false;

    //
    // What the pass that just finished decided, handed to anyone who waited for it.
    //
    private PassOutcome lastOutcome = PassOutcome.RAN;

    //
    // True once the driver has been stopped. A stopped driver runs no further passes and its loop
    // ends at the next opportunity.
    //
    private volatile boolean stopped = false;

    //
    // How long to wait before the next pass, as the last plan asked. Read by the loop after the pass
    // that set it, so the gap comes from the settings file rather than from a constant here.
    //
    private volatile long pauseMs = FALLBACK_PAUSE_MS;

    //
    // Constructs a driver over the given host, sharing the given pass lock with the import driver.
    //
    public SyncDriver(Host host, BackgroundPassLock sharedPassLock) {
        this.host = host;
        this.sharedPassLock = sharedPassLock;
    }

    //
    // Runs passes until the driver is stopped.
    //
    // A pass that fails does not end the loop. Sync has more reasons to fail than import does, not
    // fewer: no network, a remote that is briefly unreachable, an expired credential. Every one of
    // them can be gone by the next pass, and giving up would leave syncing switched on in the
    // interface and doing nothing.
    //
    public void runLoop() {
        while (!stopped) {
            try {
                runOnePass();
            }
            catch (InterruptedException interrupted) {
                Thread.currentThread().interrupt();
                return;
            }

            if (stopped) {
                return;
            }

            try {
                if (!host.pause(pauseMs)) {
                    return;
                }
            }
            catch (InterruptedException interrupted) {
                Thread.currentThread().interrupt();
                return;
            }
        }
    }

    //
    // Runs one pass, or waits for the one already running and reports what it decided.
    //
    // This is the only way a pass starts, on either platform. Two passes at once is not a state the
    // code can reach, so nothing anywhere has to stop one before starting another.
    //
    public PassOutcome runOnePass() throws InterruptedException {
        synchronized (passLock) {
            if (passRunning) {
                while (passRunning) {
                    passLock.wait();
                }
                return lastOutcome;
            }
            passRunning = true;
        }

        PassOutcome outcome = PassOutcome.RAN;
        try {
            outcome = performPass();
        }
        finally {
            synchronized (passLock) {
                passRunning = false;
                lastOutcome = outcome;
                passLock.notifyAll();
            }
        }

        return outcome;
    }

    //
    // Stops the driver: no further passes start, and a loop waiting between passes ends rather than
    // running another. A pass already in flight is left to finish; the engine pool cancels the tasks
    // it queued by source, which is what actually stops the work.
    //
    public void stop() {
        stopped = true;
    }

    //
    // True once the driver has been stopped.
    //
    public boolean isStopped() {
        return stopped;
    }

    //
    // Asks whether a sync should run and runs it.
    //
    private PassOutcome performPass() {
        SyncPlan plan;
        try {
            plan = host.readPlan();
        }
        catch (Exception error) {
            // A plan that cannot be read is not an answer of "do not sync". The pass counts as run
            // and the next one asks again.
            host.reportError("Could not work out whether a sync should run: " + error);
            return PassOutcome.RAN;
        }

        // Clamped rather than taken as given: a gap of zero or less is a loop with no gap at all, and
        // the wait it turns into on Android (Object.wait(0)) never ends.
        pauseMs = plan.pauseBetweenRunsMs > 0 ? plan.pauseBetweenRunsMs : FALLBACK_PAUSE_MS;

        if (!plan.shouldRun) {
            host.report("Not syncing: " + plan.reason);
            return PassOutcome.RAN;
        }

        if (!sharedPassLock.tryAcquire(PASS_LOCK_OWNER)) {
            // An import pass is in flight. Skipped rather than waited for: an import of a whole photo
            // library runs for the better part of an hour, and the sync it is holding up is one that
            // will be asked for again in a few minutes anyway.
            host.report("Not syncing yet: an import pass is running.");
            return PassOutcome.RAN;
        }

        try {
            host.holdAwake(true);
            try {
                host.report("Syncing \"" + plan.databasePath + "\".");

                for (SyncPlan.Step step : plan.steps) {
                    if (stopped) {
                        return PassOutcome.RAN;
                    }

                    boolean succeeded;
                    try {
                        succeeded = host.runStep(step);
                    }
                    catch (Exception error) {
                        host.reportError("Sync step \"" + step.type + "\" failed: " + error);
                        return PassOutcome.RAN;
                    }

                    if (!succeeded) {
                        // The rest of the pass is abandoned, not the loop, and the next pass starts
                        // from the plan again rather than from where this one gave up.
                        host.reportError("Sync step \"" + step.type + "\" did not succeed.");
                        return PassOutcome.RAN;
                    }
                }
            }
            finally {
                host.holdAwake(false);
            }
        }
        finally {
            sharedPassLock.release(PASS_LOCK_OWNER);
        }

        return PassOutcome.RAN;
    }
}

File: apps/ios-frontend/ios/App/App/JsEngine/SyncDriver.swift
import Foundation

//
// What one background sync pass should do, as answered by the plan-sync task.
//
// Nothing here is decided natively. The task reads the settings file, asks the platform what kind of
// connection this is, applies the same rule the app's interface applies, and hands back the task the
// pass has to run, already built. The counterpart of SyncPlan.java, deliberately identical, because
// both are answered by the same worker task.
//
struct SyncPlan {

    //
    // One task the pass runs, with its input data already serialised to the JSON string the engine
    // pool takes.
    //
    struct Step {

        //
        // The task type to queue (for example "sync-database").
        //
        let type: String

        //
        // The task's input data as a JSON string, forwarded to the engine unchanged.
        //
        let dataJson: String
    }

    //
    // Whether a sync should run right now. False never ends the loop: every reason to refuse a sync
    // can go away without the app being touched.
    //
    let shouldRun: Bool

    //
    // The sandbox-relative path of the database the pass syncs. Empty when no sync is running.
    //
    let databasePath: String

    //
    // Why no sync is running, for the log. Empty when one is.
    //
    let reason: String

    //
    // How long to wait after this pass finishes before starting the next one, in seconds.
    //
    let pauseBetweenRuns: TimeInterval

    //
    // The tasks the pass runs, in order. Empty when shouldRun is false.
    //
    let steps: [Step]
}

//
// Everything the sync driver needs that it cannot do itself: talking to the engine pool, waiting, and
// saying what happened.
//
// It is a protocol so the loop's decisions can be exercised without a running engine, and so the
// driver holds no reference to the plugin.
//
protocol SyncDriverHost: AnyObject {

    //
    // Runs the plan-sync task and returns what it says a pass should do. Throws when the task fails,
    // which the driver treats as a pass that did not work rather than as a reason to give up.
    //
    func readPlan() throws -> SyncPlan

    //
    // Runs one of the plan's steps and waits for it to finish. Returns true when it succeeded.
    //
    func runStep(_ step: SyncPlan.Step) throws -> Bool

    //
    // Waits for the given time, or until the driver is stopped. Returns false when the wait was cut
    // short by a stop, so the loop ends instead of starting another pass.
    //
    func pause(_ seconds: TimeInterval) -> Bool

    //
    // Reports what the background sync is doing.
    //
    func report(_ message: String)

    //
    // Reports something that went wrong.
    //
    func reportError(_ message: String)
}

//
// The background sync on iOS: one pass at a time, and a loop above it.
//
// This is the counterpart of SyncDriver.java, with the same single serialised entry point for running
// a pass. It matters more here than on Android, because two different things ask for a pass: the loop
// that runs while the app is foregrounded, and the background processing task the system schedules
// when it chooses. Neither knows about the other, and there is no handover to get wrong, because two
// passes at once is not a state the code can reach.
//
// Nothing a plan says ends this loop. Every reason a sync is refused can go away without the app
// being touched: a phone moves onto Wi-Fi, a network comes back, a database gets an origin, the user
// switches syncing back on. A refused pass costs a settings file read.
//
// What iOS cannot do is keep the loop running while the app is off screen. A BGProcessingTask runs
// when the system decides, typically while the phone is charging and idle, and can be killed at any
// moment, so the honest description of iOS is that it catches up in the background when the system
// allows, rather than that it syncs continuously.
//
final class SyncDriver {

    //
    // The name this driver takes the shared pass lock under, so a skipped pass can say what it is
    // waiting behind.
    //
    static let passLockOwner = "sync"

    //
    // The gap between passes when no plan has said what it should be, in seconds.
    //
    // Only reached when the very first plan read fails, because every plan carries a gap the settings
    // file resolves (see resolveSyncPauseMs in packages/api/src/lib/sync-settings.ts). It exists
    // because a gap of zero is not a gap: the loop would ask for a pass, fail to read a plan, and ask
    // again as fast as the engine could answer.
    //
    private static let fallbackPause: TimeInterval = 5 * 60

    //
    // Everything the driver needs that it cannot do itself.
    //
    private weak var host: SyncDriverHost?

    //
    // The lock that keeps a sync pass and an import pass apart.
    //
    private let sharedPassLock: BackgroundPassLock

    //
    // Guards the pass bookkeeping below, and is what a second caller waits on while a pass runs.
    //
    private let passCondition = NSCondition()

    //
    // True while a pass is in flight, so a second request waits for it rather than starting another.
    //
    private var passRunning = false

    //
    // Guards the stopped flag and the pause below, both of which are read and written from the
    // foreground loop's thread and the plugin's.
    //
    private let stateLock = NSLock()

    //
    // True once the driver has been stopped.
    //
    private var stoppedFlag = false

    //
    // How long to wait before the next pass, as the last plan asked.
    //
    private var pauseBetweenPasses: TimeInterval = SyncDriver.fallbackPause

    //
    // Constructs a driver over the given host, sharing the given pass lock with the import driver.
    //
    init(host: SyncDriverHost, sharedPassLock: BackgroundPassLock) {
        self.host = host
        self.sharedPassLock = sharedPassLock
    }

    //
    // True once the driver has been stopped.
    //
    var isStopped: Bool {
        stateLock.lock()
        defer { stateLock.unlock() }
        return stoppedFlag
    }

    //
    // Stops the driver: no further passes start, and a loop waiting between passes ends rather than
    // running another. A pass already in flight is left to finish; the engine pool cancels the tasks
    // it queued by source, which is what actually stops the work.
    //
    func stop() {
        stateLock.lock()
        stoppedFlag = true
        stateLock.unlock()
    }

    //
    // Lets the driver run again after a stop, which is what happens when the app returns to the
    // foreground.
    //
    func resume() {
        stateLock.lock()
        stoppedFlag = false
        stateLock.unlock()
    }

    //
    // Runs passes until the driver is stopped.
    //
    // A pass that fails does not end the loop. Sync has more reasons to fail than import does, not
    // fewer: no network, a remote that is briefly unreachable, an expired credential. Every one of
    // them can be gone by the next pass.
    //
    func runLoop() {
        while !isStopped {
            runOnePass()

            if isStopped {
                return
            }

            stateLock.lock()
            let waitFor = pauseBetweenPasses
            stateLock.unlock()

            guard let host = host, host.pause(waitFor) else {
                return
            }
        }
    }

    //
    // Runs one pass, or waits for the one already running.
    //
    // This is the only way a pass starts. The foreground loop asks repeatedly; the background
    // processing task asks once. Two passes at once is not a state the code can reach, so nothing
    // anywhere has to stop one before starting another.
    //
    func runOnePass() {
        passCondition.lock()
        if passRunning {
            while passRunning {
                passCondition.wait()
            }
            passCondition.unlock()
            return
        }
        passRunning = true
        passCondition.unlock()

        performPass()

        passCondition.lock()
        passRunning = false
        passCondition.broadcast()
        passCondition.unlock()
    }

    //
    // Asks whether a sync should run and runs it.
    //
    private func performPass() {
        guard let host = host else {
            return
        }

        let plan: SyncPlan
        do {
            plan = try host.readPlan()
        }
        catch {
            // A plan that cannot be read is not an answer of "do not sync". The pass counts as run
            // and the next one asks again.
            host.reportError("Could not work out whether a sync should run: \(error)")
            return
        }

        // Clamped rather than taken as given: a gap of zero or less is a loop with no gap at all.
        stateLock.lock()
        pauseBetweenPasses = plan.pauseBetweenRuns > 0 ? plan.pauseBetweenRuns : SyncDriver.fallbackPause
        stateLock.unlock()

        if !plan.shouldRun {
            host.report("Not syncing: \(plan.reason)")
            return
        }

        if !sharedPassLock.tryAcquire(SyncDriver.passLockOwner) {
            // An import pass is in flight. Skipped rather than waited for: an import of a whole photo
            // library runs for the better part of an hour, and the sync it is holding up is one that
            // will be asked for again in a few minutes anyway.
            host.report("Not syncing yet: an import pass is running.")
            return
        }

        defer { sharedPassLock.release(SyncDriver.passLockOwner) }

        host.report("Syncing \"\(plan.databasePath)\".")

        for step in plan.steps {
            if isStopped {
                return
            }

            do {
                let succeeded = try host.runStep(step)
                if !succeeded {
                    // The rest of the pass is abandoned, not the loop, and the next pass starts from
                    // the plan again rather than from where this one gave up.
                    host.reportError("Sync step \"\(step.type)\" did not succeed.")
                    return
                }
            }
            catch {
                host.reportError("Sync step \"\(step.type)\" failed: \(error)")
                return
            }
        }
    }
}
